File: src/models/Plane.js
import Passenger from './Passenger';

class Plane {
    constructor(rows, columns, speed, rowHeight, seed) {
        this.rows = rows;
        this.columns = columns;
        this.seats = Array.from({ length: rows }, () => new Array(columns).fill(null));
        this.AVERAGE_WALK_SPEED_MILES_PER_HOUR = speed;
        this.AVERAGE_WALK_SPEED_MILES_PER_SECOND = this.AVERAGE_WALK_SPEED_MILES_PER_HOUR / 3600.0;
        this.AVERAGE_WALK_SPEED_INCHES_PER_SECOND = this.AVERAGE_WALK_SPEED_MILES_PER_SECOND * 63360;
        this.AVERAGE_ROW_HEIGHT_INCHES = rowHeight;
        this.AVERAGE_ROW_HEIGHT_MILES = this.AVERAGE_ROW_HEIGHT_INCHES / 63360.0;
        this.timeSecondsToWalkPlane = this.calculateWalkTime();
        this.assemblyTimeWaveReductionFactor = 0.05;
        this.seatsHash = {}; //to let the animations look up their passengers
        this.seed = seed;
    } 

    printSeats() {
        this.seats.forEach((row, i) => {
            this.arrangeMiddleOut(row).forEach(seat => {
            });
        });
    }

    embark() {
        let count = 0;
        this.seats.forEach((row, i) => {
            row.forEach((_, j) => {
                const isAisleSeat = this.columnLabel(j) === 'C';
                const seatLabel = `${i + 1}${this.columnLabel(j)}`;
                // we can take 10 random samples inside Passenger without overlap. not sure if this is necessary or not. pretty low cost.
                let curPassenger = new Passenger(isAisleSeat, seatLabel, this.seed + (10 * count));
                this.seats[i][j] = curPassenger;
                this.seatsHash[seatLabel] = curPassenger;
                count += 1;
            });
        });

        // this.putSeatsIntoSeatsHash();
    }

    putSeatsIntoSeatsHash() {
        this.seats.forEach((row, i) => {
            row.forEach((passenger, j) => {
                const seatLabel = `${i + 1}${this.columnLabel(j)}`;

                this.seatsHash[seatLabel] = passenger;
            });
        });
    }

    columnLabel(index) {
        switch (this.columns) {
            case 1: return ['C'][index];
            case 2: return ['C', 'D'][index];
            case 3: return ['A', 'C', 'D'][index];
            case 4: return ['A', 'C', 'D', 'F'][index];
            case 5: return ['A', 'B', 'C', 'D', 'F'][index];
            case 6: return ['A', 'B', 'C', 'D', 'E', 'F'][index];
            default: return null;
        }
    }

    arrangeMiddleOut(row) {
        switch (this.columns) {
            case 1: return [row[0]];
            case 2: return [row[0], row[1]];
            case 3: return [row[0], row[1], row[2]];
            case 4: return [row[1], row[0], row[2], row[3]];
            case 5: return [row[2], row[1], row[0], row[3], row[4]];
            case 6: return [row[2], row[1], row[0], row[3], row[4], row[5]];
        }
    }

    // calculateWalkTime() {
    //     return this.rows * this.AVERAGE_ROW_HEIGHT_MILES / this.AVERAGE_WALK_SPEED_MILES_PER_SECOND;
    // }

    calculateWalkTime() {
        return this.AVERAGE_ROW_HEIGHT_INCHES * this.rows / this.AVERAGE_WALK_SPEED_INCHES_PER_SECOND;
    }
    
    disembarkCurrent() {
        let totalTime = 0;
        this.seats.forEach((row, rowIdx) => {
            const middleOutRow = this.arrangeMiddleOut(row);
            middleOutRow.forEach((passenger, idx) => {
                const assemblyTime = passenger.assemblyTimeCurrent;
                const minBuffer = passenger.minBuffer;

                // set the tracker stuff
                passenger.statusQuoTracker.seatedStart = 0;

                if (passenger.seat.indexOf('C') >= 0) {
                    passenger.statusQuoTracker.seatedEnd = 0;
                    passenger.statusQuoTracker.gatheringBelongingsStart = 0;
                } else {
                    passenger.statusQuoTracker.seatedEnd = totalTime;
                    passenger.statusQuoTracker.gatheringBelongingsStart = totalTime;
                }

                passenger.statusQuoTracker.gatheringBelongingsEnd = passenger.statusQuoTracker.gatheringBelongingsStart + passenger.assemblyTimeCurrent;

                passenger.statusQuoTracker.standingStoppedStart = passenger.statusQuoTracker.gatheringBelongingsEnd
                // the very first person off the plane does not wait minBuffer for anyone in front to leave
                if (idx === 0 && rowIdx === 0) {
                    passenger.statusQuoTracker.standingStoppedEnd = passenger.statusQuoTracker.standingStoppedStart;
                    passenger.statusQuoTracker.standingWaitingStart = passenger.statusQuoTracker.standingStoppedEnd;
                    passenger.statusQuoTracker.standingWaitingEnd = passenger.statusQuoTracker.standingWaitingStart;
                } else if (passenger.seat.indexOf('C') >= 0) { 
                    // account for the fact that the aisle has been standing this whole time, so they wait the minBuffer
                    passenger.statusQuoTracker.standingStoppedEnd = totalTime;
                    passenger.statusQuoTracker.standingWaitingStart = passenger.statusQuoTracker.standingStoppedEnd;
                    passenger.statusQuoTracker.standingWaitingEnd = passenger.statusQuoTracker.standingWaitingStart + passenger.minBuffer;
                } else {
                    passenger.statusQuoTracker.standingStoppedEnd = passenger.statusQuoTracker.standingStoppedStart;
                    passenger.statusQuoTracker.standingWaitingStart = passenger.statusQuoTracker.standingStoppedEnd;
                    // look up the passenger who deplaned before within this row, in rare cases your gather belongings step will be faster than min buffer, so if there's a delta wait until the buffer is met.

                    let precedingNeighbor = middleOutRow[idx - 1];
                    passenger.statusQuoTracker.standingWaitingEnd = passenger.statusQuoTracker.standingWaitingStart + Math.max(passenger.minBuffer - precedingNeighbor.assemblyTimeCurrent, 0);
                }
                
                passenger.statusQuoTracker.walkingStart = passenger.statusQuoTracker.standingWaitingEnd;
                passenger.statusQuoTracker.walkingEnd = passenger.statusQuoTracker.walkingStart + this.timeSecondsToWalkPlane * (rowIdx + 1) / this.rows;
                
                totalTime = passenger.statusQuoTracker.walkingStart; //if this is the last member of a row, this will be used to establish standingStoppedEnd of the first aisle member of the next row
                passenger.waitTimeSecondsCurrent = passenger.statusQuoTracker.walkingEnd;
            });
        });
        
        //must finally consider the amount of time it takes the last person on the plane to walk the plane
        totalTime += this.timeSecondsToWalkPlane;

        let maxWalkEndFromTrackers = 0;

        this.seats.forEach((row) => {
            row.forEach((passenger) => {
                // ... existing code for processing passengers ...

                // Update the maximum walk end time if the current passenger's walk end time is greater
                if (passenger.statusQuoTracker.walkingEnd > maxWalkEndFromTrackers) {
                    maxWalkEndFromTrackers = passenger.statusQuoTracker.walkingEnd;
                }
            });
        });

        return maxWalkEndFromTrackers;
    }

    disembarkFuture() {
        let totalTime = 0;

        this.aisleColumns().forEach((column, wave) => {
            const aisle = this.seats.map(row => row[this.aisleColumns().sort().indexOf(column)]);
            const reductionFactor = 1 - wave * this.assemblyTimeWaveReductionFactor;
            const maxAssemblyTime = aisle.reduce((max, passenger) => Math.max(max, passenger.assemblyTimeFuture), 0) * reductionFactor;

            aisle.forEach((passenger, idx) => {
                if (wave === 0) {
                    passenger.fillAndFlushTracker.seatedStart = 0;
                    passenger.fillAndFlushTracker.seatedEnd = 0;    
                } else {
                    passenger.fillAndFlushTracker.seatedStart = 0;
                    passenger.fillAndFlushTracker.seatedEnd = totalTime;
                }

                passenger.fillAndFlushTracker.gatheringBelongingsStart = passenger.fillAndFlushTracker.seatedEnd;
                passenger.fillAndFlushTracker.gatheringBelongingsEnd = passenger.fillAndFlushTracker.gatheringBelongingsStart + passenger.assemblyTimeFuture * reductionFactor;

                passenger.fillAndFlushTracker.standingStoppedStart = passenger.fillAndFlushTracker.gatheringBelongingsEnd;
                if (idx === 0) {
                    passenger.fillAndFlushTracker.standingStoppedEnd = passenger.fillAndFlushTracker.standingStoppedStart + (maxAssemblyTime - (passenger.assemblyTimeFuture * reductionFactor));
                    passenger.fillAndFlushTracker.standingWaitingStart = passenger.fillAndFlushTracker.standingStoppedEnd;
                    passenger.fillAndFlushTracker.standingWaitingEnd = passenger.fillAndFlushTracker.standingWaitingStart;
                } else {
                    passenger.fillAndFlushTracker.standingStoppedEnd = aisle[idx - 1].fillAndFlushTracker.walkingStart;
                    passenger.fillAndFlushTracker.standingWaitingStart = passenger.fillAndFlushTracker.standingStoppedEnd;
                    passenger.fillAndFlushTracker.standingWaitingEnd = passenger.fillAndFlushTracker.standingWaitingStart + passenger.minBuffer;
                }

                passenger.fillAndFlushTracker.walkingStart = passenger.fillAndFlushTracker.standingWaitingEnd;
                passenger.fillAndFlushTracker.walkingEnd = passenger.fillAndFlushTracker.walkingStart + this.timeSecondsToWalkPlane * (idx + 1) / this.rows;          
                passenger.waitTimeSecondsFuture = passenger.fillAndFlushTracker.walkingEnd
            });

            totalTime = aisle[this.rows - 1].fillAndFlushTracker.walkingEnd;
        });

        let maxWalkEndFromTrackers = 0;
        this.seats.forEach((row) => {
            row.forEach((passenger) => {
                // ... existing code for processing passengers ...

                // Update the maximum walk end time if the current passenger's walk end time is greater
                if (passenger.fillAndFlushTracker.walkingEnd > maxWalkEndFromTrackers) {
                    maxWalkEndFromTrackers = passenger.fillAndFlushTracker.walkingEnd;
                }
            });
        });

        return maxWalkEndFromTrackers;
    }

    aisleColumns() {
        switch (this.columns) {
            case 1: return ['C'];
            case 2: return ['C', 'D'];
            case 3: return ['C', 'A', 'D'];
            case 4: return ['C', 'D', 'A', 'F'];
            case 5: return ['C', 'D', 'B', 'F', 'A'];
            case 6: return ['C', 'D', 'B', 'E', 'A', 'F'];
            default: return [];
        }
    }
}

export default Plane;

File: original_ruby_stuff/plane.rb
require_relative 'passenger'
require 'byebug'
require 'awesome_print'
require 'set'

class Plane
  attr_reader :rows, :columns, :seats, :time_seconds_to_walk_plane
  
  AVERAGE_ROW_HEIGHT_INCHES = 31
  AVERAGE_ROW_HEIGHT_MILES =  AVERAGE_ROW_HEIGHT_INCHES / 63360.0
  AVERAGE_WALK_SPEED_MILES_PER_HOUR = 3.1
  AVERAGE_WALK_SPEED_MILES_PER_SECOND = AVERAGE_WALK_SPEED_MILES_PER_HOUR / 3600.0

  def initialize(rows, columns)
    @rows = rows
    @columns = columns
    @aisle_seats = ['C', 'D']
    @seats = Array.new(rows) { Array.new(columns) }
    @time_seconds_to_walk_plane = calculate_walk_time
  end
  
  def print_seats
    @seats.each_with_index do |row, i|
      arrange_middle_out(row).each do |seat|
        puts "#{i+1}#{column_label(row.index(seat))}"
      end
    end
  end

  def embark
    # Fill the plane with passengers
    @seats.each_with_index do |row, i|
      row.each_with_index do |seat, j|
        is_aisle_seat = column_label(j) == 'C'  # Only 'C' side gets the aisle bonus
        seat_label = "#{i+1}#{column_label(j)}"
        @seats[i][j] = Passenger.new(is_aisle_seat, seat_label)
      end
    end
  end

  def disembark_current
    # Simulate deplaning in current system
    total_time = 0
    @seats.each do |row|
      middle_out_row = arrange_middle_out(row)
      middle_out_row.each_with_index do |passenger, idx|
        # if youre the first person in the middle out row, i.e. you're the (1 / width) who gets to stand, the time impact is slightly different
        if idx == 0
          # handle case where you're at front of plane:
          if total_time < passenger.assembly_time_current
            if total_time == 0
              time = passenger.assembly_time_current - total_time
            else
              time = [passenger.assembly_time_current - total_time, passenger.min_buffer].max
            end
          else
            assembly_time = passenger.assembly_time_current
            min_buffer = passenger.min_buffer
            time = [min_buffer].max
          end
        else
          assembly_time = passenger.assembly_time_current
          min_buffer = passenger.min_buffer
          time = [assembly_time, min_buffer].max          
        end
        
        total_time += time

        passenger.wait_time_seconds_current = total_time
      end
    end
    total_time
  end
    
  def disembark_future
    # Simulate deplaning in new system
    total_time = 0
    
    aisle_columns.each_with_index do |column, wave|
      aisle = @seats.map { |row| row[aisle_columns.sort.index(column)] }
      
      reduction_factor = 1 - wave * 0.05  # 5% reduction in assembly time for each new wave, because easier when fewer other people on plane
      max_assembly_time = aisle.max_by(&:assembly_time_future).assembly_time_future * reduction_factor
      total_time += max_assembly_time
      # puts aisle
      # ap"disembarking aisle: #{aisle}"
      
      aisle.each_with_index do |passenger, idx|
        #calculate the time it takes the PASSENGER to get off here
        if idx == 0 
          # if at the front of the plane, no buffer, so the time they wait is the total_time
          passenger.wait_time_seconds_future = total_time
        else
          # if not at front, they wait whatever their neighbor waited PLUS the min_buffer
          passenger.wait_time_seconds_future = aisle[idx - 1].wait_time_seconds_future + passenger.min_buffer
        end
        
        total_time += passenger.min_buffer
      end
      
      total_time += @time_seconds_to_walk_plane
    end
    total_time
  end

  private
  
  def aisle_columns
    # TODO: what sequence of column-based deplaning allows for groups to stick together best?
    # this is the actual order of flushing in each scenario - so flush c, then d, then a, then f in the 4 case
    case @columns
    when 1
      ['C']
    when 2
      ['C', 'D']
    when 3
      ['C', 'A', 'D']
    when 4
      ['C', 'D', 'A', 'F']
    when 5
      ['C', 'D', 'B', 'F', 'A']
    when 6
      ['C', 'D', 'B', 'E', 'A', 'F']
    end
  end
  
  def column_label(index)
    ('A'..'F').to_a[index]
  end
  
  def arrange_middle_out(row)
    if @columns == 1 #C
      [row[0]]
    elsif @columns == 2 #CD
      [row[0], row[1]]
    elsif @columns == 3 #CAD
      [row[0], row[1], row[2]]
    elsif @columns == 4
      [row[1], row[0], row[2], row[3]]  # CADF
    elsif @columns == 5
      [row[2], row[1], row[0], row[3], row[4]]  # CBADF
    elsif @columns == 6
      [row[2], row[1], row[0], row[3], row[4], row[5]]  # CBADEF
    end
  end
  
  def calculate_walk_time
    @rows * AVERAGE_ROW_HEIGHT_MILES / AVERAGE_WALK_SPEED_MILES_PER_SECOND
  end
end